### DISCUSSION TITLE
Will `unstable_cache` be stabilized alongside `use cache`?

### DISCUSSION THREAD

**Discussion by sannajammeh** [vercel/next.js#71835] [App Router]
`use cache` is an excellent addition to the Next.js codebase, however it does not fit in with libraries as its so tightly coupled to the Next.js codebase. 

We have several libraries in the works for CMS's which are able to be provided a cacheFunction when initialized. This makes the library fully able to run both in Next.js and in other frameworks. We then leverage the heuristics of `unstable_cache` in our project code to enable a full cache on the library.

Take the following user code example:

`src/lib/animal-api.ts`
```tsx
import { unstable_cache as cache } from "next/cache";
import { createAnimalApi, type CacheFn } from "animal-registry";

export const cacheFunction: CacheFn = (fn, { keyParts, tags, preview }) => {
  if (preview) {
    noStore();
    return fn;
  }
  return cache(fn, keyParts, { tags }); // We have now enabled permanent caching and cache tagging based on the library's output
};

export const animalApi = createAnimalApi({
	accessToken: "...",
    cacheFunction,
})
```

`src/app/page.tsx`
```tsx
export default async function Page() {
  // this calls the unstable_cache fn with the correct serialized args, tags etc.
  const cats = await animalApi.getCats({limit: 4}) // tagged with `animal:cats` - cache key consists of limit: 4 and stringified internal fn
  const dogs = await animalApi.getDocs({limit: 3}) // tagged with `animal:dogs`; cache key consists of limit: 3 and stringified internal fn

 // page.tsx is now permanently cached and tagged ith animal:cats & animal:dogs which we can revalidate on later
 return <>{JSON.stringify(cats)}</>
}
``` 

Afaik this behaviour which works excellent for us in our current libraries is not possible with the new "use cache" API. 

I've made an initial attempt which does not work:
```tsx
type Callback = (...args: any[]) => Promise<any>;

const cacheFnFactory = <T extends Callback>(
  callback: T,
  keyParts: string[],
  tags: string[]
): T => {
  const cacheFn = async (args: Parameters<T>, tags: string[], _: string[]) => {
    "use cache";
    const data = await callback(args);
    for (const tag of tags) {
      cacheTag(tag);
    }
    return data;
  };

  return ((...args: Parameters<T>) => cacheFn(args, tags, keyParts)) as T;
};

const DynamicComponent = async () => {
  const cachedGetResult = cacheFnFactory(
    (data: string) => setTimeout(500, <p>{data}</p>),
    ["some-unique-key"],
    ["tagged"]
  );

  console.log(await cachedGetResult("Hello world"));
  return null;
};
```

This results in the following error:

```bash
[ Server ] Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server". Or maybe you meant to call this function rather than return it.
  [function]
```

I sincerely hope you either A. allow an outer scope to provide a function into a "use cache" function, or B keep unstable_cache around and stabilize it alongside "use cache" options.

For option A, I don't really see an issue allowing the outer scoped `callback` to be accessible inside the inner function as it will only be called during a cache miss/stale, meaning the `callback` function is already fresh.

**Comment by tomdohnal** [comment]
As per [the docs](https://nextjs.org/docs/canary/app/api-reference/directives/use-cache), the `"use cache"` directive is aimed to replace the `unstable_cache` function:"use cache"` directive is aimed to replace the `unstable_cache` function:
>The use cache directive is an experimental feature that aims to replace the [unstable_cache](https://nextjs.org/docs/canary/app/api-reference/legacy-apis/unstable_cache) function. Unlike unstable_cache, which is limited to caching JSON data and requires manual definition of revalidation periods and tags, use cache offers more flexibility. It allows you to cache a wider range of data, including anything that React Server Components (RSC) can serialize, as well as data-fetching outputs and component outputs.

**Comment by romeovs** [comment]
I am also building a library and set of tools that does caching at several layers that the end-user does not have access too. Currently the library allows users to drop in any cache function so it's super-easy to integrate with Next.js:

```
import { buildExample } from "example-lib"
import { unstable_cache } from "next/cache"

const example = buildExample({
  cache: unstable_cache,
})
```

It's unclear how this will work with `"use cache"`, so I hope it will be possible to either keep using `unstable_cache`, or be able to implement it manually on top of `"use cache"`.
